File: archive-application/src/main/java/site/archive/service/user/UserRegisterServiceV1.java
package site.archive.service.user;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import site.archive.common.exception.common.DuplicateResourceException;
import site.archive.domain.user.BaseUser;
import site.archive.domain.user.UserInfo;
import site.archive.domain.user.UserRepository;
import site.archive.dto.v1.auth.BasicRegisterCommandV1;
import site.archive.dto.v1.auth.OAuthRegisterCommandV1;
import site.archive.dto.v1.user.BaseUserDto;
import site.archive.service.message.MessagingService;

@Service
@Transactional
@Slf4j
@RequiredArgsConstructor
public class UserRegisterServiceV1 {

    private final UserRepository userRepository;
    private final MessagingService messagingService;

    public long getOrRegisterUser(BasicRegisterCommandV1 registerCommand) {
        var user = userRepository.findByMailAddress(registerCommand.getEmail())
                                 .orElseGet(() -> registerUser(registerCommand));
        return user.getId();
    }

    public UserInfo getOrRegisterUserReturnInfo(BasicRegisterCommandV1 registerCommand) {
        var user = userRepository.findByMailAddress(registerCommand.getEmail())
                                 .orElseGet(() -> registerUser(registerCommand));
        return user.convertToUserInfo();
    }

    public BaseUser registerUser(BasicRegisterCommandV1 registerCommand) {
        try {
            var user = userRepository.save(registerCommand.toUserEntity());
            sendRegisterNotification(registerCommand, user);
            return user;
        } catch (DataIntegrityViolationException e) {
            throw new DuplicateResourceException("이메일이 이미 존재합니다.");
        }
    }

    private void sendRegisterNotification(BasicRegisterCommandV1 registerCommand, BaseUser user) {
        if (registerCommand instanceof OAuthRegisterCommandV1 oAuthRegisterCommand) {
            var oauthProvider = oAuthRegisterCommand.getProvider().getRegistrationId();
            messagingService.sendUserRegisterMessage(BaseUserDto.from(user), oauthProvider);
        } else {
            messagingService.sendUserRegisterMessage(BaseUserDto.from(user), "Id/Password");
        }
    }

}

File: archive-application/src/main/kotlin/site/archive/service/user/UserRegisterServiceV1.kt
package site.archive.service.user

import org.slf4j.LoggerFactory
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import site.archive.common.exception.common.DuplicateResourceException
import site.archive.domain.user.BaseUser
import site.archive.domain.user.PasswordUser
import site.archive.domain.user.UserInfo
import site.archive.domain.user.UserRepository
import site.archive.dto.v1.auth.BasicRegisterCommandV1
import site.archive.dto.v1.auth.OAuthRegisterCommandV1
import site.archive.dto.v1.user.BaseUserDtoV1
import site.archive.service.message.MessagingService

@Service
@Transactional(readOnly = true)
class UserRegisterServiceV1(val userRepository: UserRepository, val messagingService: MessagingService) {

    private val log = LoggerFactory.getLogger(javaClass)


    @Transactional
    fun getOrRegisterUser(registerCommand: BasicRegisterCommandV1): Long {
        val user = userRepository.findByMailAddress(registerCommand.email)
            .orElseGet { registerUser(registerCommand) }
        return user.id
    }

    @Transactional
    fun getOrRegisterUserReturnInfo(registerCommand: BasicRegisterCommandV1): UserInfo {
        val user = userRepository.findByMailAddress(registerCommand.email)
            .orElseGet { registerUser(registerCommand) }
        return user.convertToUserInfo()
    }

    @Transactional
    fun registerUser(registerCommand: BasicRegisterCommandV1): BaseUser {
        try {
            val user = userRepository.save(registerCommand.toUserEntity())
            sendRegisterNotification(registerCommand, user)
            return user
        } catch (e: DataIntegrityViolationException) {
            log.error("[v1] 이메일이 이미 존재합니다.", e)
            throw DuplicateResourceException("이메일이 이미 존재합니다.")
        }
    }

    private fun sendRegisterNotification(registerCommand: BasicRegisterCommandV1, user: BaseUser) {
        if (registerCommand is OAuthRegisterCommandV1) {
            val oauthProvider = registerCommand.provider.registrationId
            messagingService.sendUserRegisterMessage(BaseUserDtoV1.from(user), oauthProvider)
        } else {
            messagingService.sendUserRegisterMessage(BaseUserDtoV1.from(user), PasswordUser.PASSWORD_TYPE)
        }
    }

}